risingwavelabs/risingwave
# feat(streaming): add single-row cache for monotonically increasing dynamic filter

I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

### DynamicFilterCache

It only works for `LHS > NOW()` filter conditions. 
- [ ] Support >= NOW also.
- [ ] Assume RHS is monotonically increasing, not strictly increasing.
- [ ] Handle state cleaning with watermark.

Caches a single row that is larger than the filter value, and is the minimal value. Use `Match(Row)` to do so.
If no such row exists, it means all value are smaller than the filter value (current NOW()).
In that case we should also indicate it, so we can avoid table scan. Use `NoMatch` to do so.

We update the cache at 3 occasions:
1. Processing LHS chunk. For insert / updateinsert, we may need to refresh the cached value. For Delete/UpdateDelete, we may need to evict the cache.
2. Processing RHS chunk. Similar to step 1, cases are listed in the implementation.
3. Processing table scan. We will also need to refresh cache here. For instance in RHS / LHS processing, cache may be evicted. In that case, it can only be refreshed in table scan, since we don't know previous values in cache.

### Testing Scenario

1. Create a source stream with LHS values lagging NOW(), i.e. no match.
2. Create dynamic filter query with the source stream on LHS and run it.
3. Observe barrier latency.
4. Observe e2e throughput, make sure no regression.
5. Observe reads.

### TODOs

- [x] Add design of cache (as a docstring)
- [x] Implement it in dynamic filter.
- [x] Add binary comparator function. We need it to update / evict cache.
- [ ] Test it by observing dynamic filter reads, they should decrease.
- [ ] Performance should be the same.
- [ ] Add optimizer check to decide if we can have cache. Only support `> NOW()` for now.



## Checklist

- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.



<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note



</details>
